README.md
<p align="center"><img src="https://horizom.github.io/img/horizom-logo-color.svg" width="400"></p>

<p align="center">
<a href="https://packagist.org/packages/horizom/auth"><img src="https://poser.pugx.org/horizom/auth/d/total.svg" alt="Total Downloads"></a>
<a href="https://packagist.org/packages/horizom/auth"><img src="https://poser.pugx.org/horizom/auth/v/stable.svg" alt="Latest Stable Version"></a>
<a href="https://packagist.org/packages/horizom/auth"><img src="https://poser.pugx.org/horizom/auth/license.svg" alt="License"></a>
</p>

# Horizom\Auth

Authentication for PHP. Simple, lightweight and secure. Written once, to be used everywhere. Completely framework-agnostic and database-agnostic.

[Full documentation](https://horizom.github.io/docs/framework/Authentication)

## Why do I need this?

* There are [tons](https://www.troyhunt.com/whos-who-of-bad-password-practices/) [of](https://blog.codinghorror.com/password-rules-are-bullshit/) [websites](https://badpasswordpolicies.tumblr.com/) with weak authentication systems. Don’t build such a site.
* Re-implementing a new authentication system for every PHP project is *not* a good idea.
* Building your own authentication classes piece by piece, and copying it to every project, is *not* recommended, either.
* A secure authentication system with an easy-to-use API should be thoroughly designed and planned.
* Peer-review for your critical infrastructure is *a must*.

## Requirements

* PHP 5.6.0+
  * PDO (PHP Data Objects) extension (`pdo`)
    * MySQL Native Driver (`mysqlnd`) **or** PostgreSQL driver (`pgsql`) **or** SQLite driver (`sqlite`)
  * OpenSSL extension (`openssl`)
* MySQL 5.5.3+ **or** MariaDB 5.5.23+ **or** PostgreSQL 9.5.10+ **or** SQLite 3.14.1+ **or** [other SQL databases](Database)

## Installation

 1. Include the library via Composer :

    ```
    composer require horizom/auth
    ```

 2. Include the Composer autoloader:

    ```php
    require __DIR__ . '/vendor/autoload.php';
    ```

 3. Set up a database and create the required tables:

    * [MariaDB](Database/MySQL.sql)
    * [MySQL](Database/MySQL.sql)
    * [PostgreSQL](Database/PostgreSQL.sql)
    * [SQLite](Database/SQLite.sql)

## Upgrading

Migrating from an earlier version of this project? See our [upgrade guide](Migration.md) for help.

## Usage

- [Horizom\Auth](#horizomauth)
  - [Why do I need this?](#why-do-i-need-this)
  - [Requirements](#requirements)
  - [Installation](#installation)
  - [Upgrading](#upgrading)
  - [Usage](#usage)
    - [Creating a new instance](#creating-a-new-instance)
    - [Registration (sign up)](#registration-sign-up)
    - [Login (sign in)](#login-sign-in)
    - [Email verification](#email-verification)
    - [Keeping the user logged in](#keeping-the-user-logged-in)
    - [Password reset (“forgot password”)](#password-reset-forgot-password)
      - [Step 1 of 3: Initiating the request](#step-1-of-3-initiating-the-request)
      - [Step 2 of 3: Verifying an attempt](#step-2-of-3-verifying-an-attempt)
      - [Step 3 of 3: Updating the password](#step-3-of-3-updating-the-password)
    - [Changing the current user’s password](#changing-the-current-users-password)
    - [Changing the current user’s email address](#changing-the-current-users-email-address)
    - [Re-sending confirmation requests](#re-sending-confirmation-requests)
    - [Logout](#logout)
    - [Accessing user information](#accessing-user-information)
      - [Login state](#login-state)
      - [User ID](#user-id)
      - [Email address](#email-address)
      - [Display name](#display-name)
      - [Status information](#status-information)
      - [Checking whether the user was “remembered”](#checking-whether-the-user-was-remembered)
      - [IP address](#ip-address)
      - [Additional user information](#additional-user-information)
    - [Reconfirming the user’s password](#reconfirming-the-users-password)
    - [Roles (or groups)](#roles-or-groups)
      - [Checking roles](#checking-roles)
      - [Available roles](#available-roles)
      - [Permissions (or access rights, privileges or capabilities)](#permissions-or-access-rights-privileges-or-capabilities)
      - [Custom role names](#custom-role-names)
    - [Enabling or disabling password resets](#enabling-or-disabling-password-resets)
    - [Throttling or rate limiting](#throttling-or-rate-limiting)
    - [Administration (managing users)](#administration-managing-users)
      - [Creating new users](#creating-new-users)
      - [Deleting users](#deleting-users)
      - [Retrieving a list of registered users](#retrieving-a-list-of-registered-users)
      - [Assigning roles to users](#assigning-roles-to-users)
      - [Taking roles away from users](#taking-roles-away-from-users)
      - [Checking roles](#checking-roles-1)
      - [Impersonating users (logging in as user)](#impersonating-users-logging-in-as-user)
      - [Changing a user’s password](#changing-a-users-password)
    - [Cookies](#cookies)
      - [Renaming the library’s cookies](#renaming-the-librarys-cookies)
      - [Defining the domain scope for cookies](#defining-the-domain-scope-for-cookies)
      - [Restricting the path where cookies are available](#restricting-the-path-where-cookies-are-available)
      - [Controlling client-side script access to cookies](#controlling-client-side-script-access-to-cookies)
      - [Configuring transport security for cookies](#configuring-transport-security-for-cookies)
    - [Utilities](#utilities)
      - [Creating a random string](#creating-a-random-string)
      - [Creating a UUID v4 as per RFC 4122](#creating-a-uuid-v4-as-per-rfc-4122)
    - [Reading and writing session data](#reading-and-writing-session-data)
  - [Frequently asked questions](#frequently-asked-questions)
    - [What about password hashing?](#what-about-password-hashing)
    - [How can I implement custom password requirements?](#how-can-i-implement-custom-password-requirements)
    - [Why are there problems when using other libraries that work with sessions?](#why-are-there-problems-when-using-other-libraries-that-work-with-sessions)
    - [Why are other sites not able to frame or embed my site?](#why-are-other-sites-not-able-to-frame-or-embed-my-site)
  - [Exceptions](#exceptions)
  - [General advice](#general-advice)
  - [Contributing](#contributing)
  - [License](#license)

### Creating a new instance

```php
$db = new \PDO('mysql:dbname=my-database;host=localhost;charset=utf8mb4', 'my-username', 'my-password');
// or
$db = new \PDO('pgsql:dbname=my-database;host=localhost;port=5432', 'my-username', 'my-password');
// or
$db = new \PDO('sqlite:../Databases/my-database.sqlite');

$auth = new \Horizom\Auth\Auth($db);

// Set password hashing algorithm
$auth->setPasswordHashAlgorithm(PASSWORD_DEFAULT);
```

If you have an open `PDO` connection already, just re-use it. The database user (e.g. `my-username`) needs at least the privileges `SELECT`, `INSERT`, `UPDATE` and `DELETE` for the tables used by this library (or their parent database).

If your web server is behind a proxy server and `$_SERVER['REMOTE_ADDR']` only contains the proxy’s IP address, you must pass the user’s real IP address to the constructor in the second argument, which is named `$ipAddress`. The default is the usual remote IP address received by PHP.

Should your database tables for this library need a common prefix, e.g. `my_users` instead of `users` (and likewise for the other tables), pass the prefix (e.g. `my_`) as the third parameter to the constructor, which is named `$dbTablePrefix`. This is optional and the prefix is empty by default.

During development, you may want to disable the request limiting or throttling performed by this library. To do so, pass `false` to the constructor as the fourth argument, which is named `$throttling`. The feature is enabled by default.

During the lifetime of a session, some user data may be changed remotely, either by a client in another session or by an administrator. That means this information must be regularly resynchronized with its authoritative source in the database, which this library does automatically. By default, this happens every five minutes. If you want to change this interval, pass a custom interval in seconds to the constructor as the fifth argument, which is named `$sessionResyncInterval`.

If all your database tables need a common database name, schema name, or other qualifier that must be specified explicitly, you can optionally pass that qualifier to the constructor as the sixth parameter, which is named `$dbSchema`.

### Registration (sign up)

```php
try {
    $userId = $auth->register($_POST['email'], $_POST['password'], $_POST['username'], function ($selector, $token) {
        echo 'Send ' . $selector . ' and ' . $token . ' to the user (e.g. via email)';
    });

    echo 'We have signed up a new user with the ID ' . $userId;
}
catch (\Horizom\Auth\Exception\InvalidEmailException $e) {
    die('Invalid email address');
}
catch (\Horizom\Auth\Exception\InvalidPasswordException $e) {
    die('Invalid password');
}
catch (\Horizom\Auth\Exception\UserAlreadyExistsException $e) {
    die('User already exists');
}
catch (\Horizom\Auth\Exception\TooManyRequestsException $e) {
    die('Too many requests');
}
```

**Note:** The anonymous callback function is a [closure](https://www.php.net/manual/functions.anonymous.php). Thus, besides its own parameters, only [superglobals](https://www.php.net/manual/language.variables.superglobals.php) like `$_GET`, `$_POST`, `$_COOKIE` and `$_SERVER` are available inside. For any other variable from the parent scope, you need to explicitly make a copy available inside by adding a `use` clause after the parameter list.

The username in the third parameter is optional. You can pass `null` there if you don’t want to manage usernames.

If you want to enforce unique usernames, on the other hand, simply call `registerWithUniqueUsername` instead of `register`, and be prepared to catch the `DuplicateUsernameException`.

**Note:** When accepting and managing usernames, you may want to exclude non-printing control characters and certain printable special characters, as in the character class `[\x00-\x1f\x7f\/:\\]`. In order to do so, you could wrap the call to `Auth#register` or `Auth#registerWithUniqueUsername` inside a conditional branch, for example by only accepting usernames when the following condition is satisfied:

```php
if (\preg_match('/[\x00-\x1f\x7f\/:\\\\]/', $username) === 0) {
    // ...
}
```

For email verification, you should build an URL with the selector and token and send it to the user, e.g.:

```php
$url = 'https://www.example.com/verify_email?selector=' . \urlencode($selector) . '&token=' . \urlencode($token);
```

If you don’t want to perform email verification, just omit the last parameter to `Auth#register`. The new user will be active immediately, then.

Need to store additional user information? Read on [here](#additional-user-information).

**Note:** When sending an email to the user, please note that the (optional) username, at this point, has not yet been confirmed as acceptable to the owner of the (new) email address. It could contain offensive or misleading language chosen by someone who is not actually the owner of the address.

### Login (sign in)

```php
try {
    $auth->login($_POST['email'], $_POST['password']);

    echo 'User is logged in';
}
catch (\Horizom\Auth\Exception\InvalidEmailException $e) {
    die('Wrong email address');
}
catch (\Horizom\Auth\Exception\InvalidPasswordException $e) {
    die('Wrong password');
}
catch (\Horizom\Auth\Exception\EmailNotVerifiedException $e) {
    die('Email not verified');
}
catch (\Horizom\Auth\Exception\TooManyRequestsException $e) {
    die('Too many requests');
}
```

If you want to sign in with usernames on the other hand, either in addition to the login via email address or as a replacement, that’s possible as well. Simply call the method `loginWithUsername` instead of method `login`. Then, instead of catching `InvalidEmailException`, make sure to catch both `UnknownUsernameException` and `AmbiguousUsernameException`. You may also want to read the notes about the uniqueness of usernames in the section that explains how to [sign up new users](#registration-sign-up).

### Email verification

Extract the selector and token from the URL that the user clicked on in the verification email.

```php
try {
    $auth->confirmEmail($_GET['selector'], $_GET['token']);

    echo 'Email address has been verified';
}
catch (\Horizom\Auth\Exception\InvalidSelectorTokenPairException $e) {
    die('Invalid token');
}
catch (\Horizom\Auth\Exception\TokenExpiredException $e) {
    die('Token expired');
}
catch (\Horizom\Auth\Exception\UserAlreadyExistsException $e) {
    die('Email address already exists');
}
catch (\Horizom\Auth\Exception\TooManyRequestsException $e) {
    die('Too many requests');
}
```

If you want the user to be automatically signed in after successful confirmation, just call `confirmEmailAndSignIn` instead of `confirmEmail`. That alternative method also supports [persistent logins](#keeping-the-user-logged-in) via its optional third parameter.

On success, the two methods `confirmEmail` and `confirmEmailAndSignIn` both return an array with the user’s new email address, which has just been verified, at index one. If the confirmation was for an address change instead of a simple address verification, the user’s old email address will be included in the array at index zero.

### Keeping the user logged in

The third parameter to the `Auth#login` and `Auth#confirmEmailAndSignIn` methods controls whether the login is persistent with a long-lived cookie. With such a persistent login, users may stay authenticated for a long time, even when the browser session has already been closed and the session cookies have expired. Typically, you’ll want to keep the user logged in for weeks or months with this feature, which is known as “remember me” or “keep me logged in”. Many users will find this more convenient, but it may be less secure if they leave their devices unattended.

```php
if ($_POST['remember'] == 1) {
    // keep logged in for one year
    $rememberDuration = (int) (60 * 60 * 24 * 365.25);
}
else {
    // do not keep logged in after session ends
    $rememberDuration = null;
}

// ...

$auth->login($_POST['email'], $_POST['password'], $rememberDuration);

// ...
```

*Without* the persistent login, which is the *default* behavior, a user will only stay logged in until they close their browser, or as long as configured via `session.cookie_lifetime` and `session.gc_maxlifetime` in PHP.

Omit the third parameter or set it to `null` to disable the feature. Otherwise, you may ask the user whether they want to enable “remember me”. This is usually done with a checkbox in your user interface. Use the input from that checkbox to decide between `null` and a pre-defined duration in seconds here, e.g. `60 * 60 * 24 * 365.25` for one year.

### Password reset (“forgot password”)

#### Step 1 of 3: Initiating the request

```php
try {
    $auth->forgotPassword($_POST['email'], function ($selector, $token) {
        echo 'Send ' . $selector . ' and ' . $token . ' to the user (e.g. via email)';
    });

    echo 'Request has been generated';
}
catch (\Horizom\Auth\Exception\InvalidEmailException $e) {
    die('Invalid email address');
}
catch (\Horizom\Auth\Exception\EmailNotVerifiedException $e) {
    die('Email not verified');
}
catch (\Horizom\Auth\Exception\ResetDisabledException $e) {
    die('Password reset is disabled');
}
catch (\Horizom\Auth\Exception\TooManyRequestsException $e) {
    die('Too many requests');
}
```

**Note:** The anonymous callback function is a [closure](https://www.php.net/manual/functions.anonymous.php). Thus, besides its own parameters, only [superglobals](https://www.php.net/manual/language.variables.superglobals.php) like `$_GET`, `$_POST`, `$_COOKIE` and `$_SERVER` are available inside. For any other variable from the parent scope, you need to explicitly make a copy available inside by adding a `use` clause after the parameter list.

You should build an URL with the selector and token and send it to the user, e.g.:

```php
$url = 'https://www.example.com/reset_password?selector=' . \urlencode($selector) . '&token=' . \urlencode($token);
```


CHANGELOG.md
# CHANGELOG

## Commit 13-03-2022 (1.0.2)

- Rename `UserManager` class to `AuthManager`.
- Fix bug : property visibility
  - `AuthManager::$table` => `protected`.
  - `AuthManager::$passwordHashAlgo` => `protected`.
  
## Commit 04-03-2022 (1.0.1)

- Add method `setPasswordHashAlgorithm()` in `Horizom\Auth\UserManager` class to set the password algorithm. The default algorithm is `PASSWORD_DEFAULT`.

## Commit 21-09-2021 (1.0.0)

- Initial release


src/UserManager.php
<?php

namespace Horizom\Auth;

use Delight\Base64\Base64;
use Delight\Cookie\Session;
use Delight\Db\PdoDatabase;
use Delight\Db\PdoDsn;
use Delight\Db\Throwable\Error;
use Delight\Db\Throwable\IntegrityConstraintViolationException;
use Horizom\Auth\Error\DatabaseError;
use Horizom\Auth\Error\MissingCallbackError;
use Horizom\Auth\Exception\UnknownIdException;
use Horizom\Auth\Exception\AmbiguousUsernameException;
use Horizom\Auth\Exception\InvalidPasswordException;
use Horizom\Auth\Exception\UnknownUsernameException;
use Horizom\Auth\Exception\InvalidEmailException;
use Horizom\Auth\Exception\UserAlreadyExistsException;
use Horizom\Auth\Exception\DuplicateUsernameException;

/**
 * Abstract base class for components implementing user management
 *
 * @internal
 */
abstract class UserManager
{

    /** @var string session field for whether the client is currently signed in */
    const SESSION_FIELD_LOGGED_IN = 'auth_logged_in';
    /** @var string session field for the ID of the user who is currently signed in (if any) */
    const SESSION_FIELD_USER_ID = 'auth_user_id';
    /** @var string session field for the email address of the user who is currently signed in (if any) */
    const SESSION_FIELD_EMAIL = 'auth_email';
    /** @var string session field for the display name (if any) of the user who is currently signed in (if any) */
    const SESSION_FIELD_USERNAME = 'auth_username';
    /** @var string session field for the status of the user who is currently signed in (if any) as one of the constants from the {@see Status} class */
    const SESSION_FIELD_STATUS = 'auth_status';
    /** @var string session field for the roles of the user who is currently signed in (if any) as a bitmask using constants from the {@see Role} class */
    const SESSION_FIELD_ROLES = 'auth_roles';
    /** @var string session field for whether the user who is currently signed in (if any) has been remembered (instead of them having authenticated actively) */
    const SESSION_FIELD_REMEMBERED = 'auth_remembered';
    /** @var string session field for the UNIX timestamp in seconds of the session data's last resynchronization with its authoritative source in the database */
    const SESSION_FIELD_LAST_RESYNC = 'auth_last_resync';
    /** @var string session field for the counter that keeps track of forced logouts that need to be performed in the current session */
    const SESSION_FIELD_FORCE_LOGOUT = 'auth_force_logout';

    /** @var PdoDatabase the database connection to operate on */
    protected $db;
    /** @var string|null the schema name for all database tables used by this component */
    protected $dbSchema;
    /** @var string the prefix for the names of all database tables used by this component */
    protected $dbTablePrefix;
    /** @var array auth database table stack */
    private static $tables = [
        'users' => 'users',
        'confirmations' => 'users_confirmations',
        'remembered' => 'users_remembered',
        'resets' => 'users_resets',
        'throttling' => 'users_throttling',
    ];

    /** @var string password hash algorithm */
    private $passwordHashAlgo = PASSWORD_DEFAULT;

    /**
     * @param PdoDatabase|PdoDsn|\PDO $databaseConnection the database connection to operate on
     * @param string|null $dbTablePrefix (optional) the prefix for the names of all database tables used by this component
     * @param string|null $dbSchema (optional) the schema name for all database tables used by this component
     */
    protected function __construct($databaseConnection, array $tables = null, $dbTablePrefix = null, $dbSchema = null)
    {
        if ($databaseConnection instanceof PdoDatabase) {
            $this->db = $databaseConnection;
        } elseif ($databaseConnection instanceof PdoDsn) {
            $this->db = PdoDatabase::fromDsn($databaseConnection);
        } elseif ($databaseConnection instanceof \PDO) {
            $this->db = PdoDatabase::fromPdo($databaseConnection, true);
        } else {
            $this->db = null;

            throw new \InvalidArgumentException('The database connection must be an instance of either `PdoDatabase`, `PdoDsn` or `PDO`');
        }

        if ($tables !== null) {
            self::$tables = $tables;
        }

        $this->dbSchema = $dbSchema !== null ? (string) $dbSchema : null;
        $this->dbTablePrefix = (string) $dbTablePrefix;
    }

    /**
     * Set password hash algorithm
     */
    public function setPasswordHashAlgorithm(string $algorithm)
    {
        $this->passwordHashAlgo = $algorithm;
    }

    /**
     * Creates a random string with the given maximum length
     *
     * With the default parameter, the output should contain at least as much randomness as a UUID
     *
     * @param int $maxLength the maximum length of the output string (integer multiple of 4)
     * @return string the new random string
     */
    public static function createRandomString($maxLength = 24)
    {
        // calculate how many bytes of randomness we need for the specified string length
        $bytes = \floor((int) $maxLength / 4) * 3;

        // get random data
        $data = \openssl_random_pseudo_bytes($bytes);

        // return the Base64-encoded result
        return Base64::encodeUrlSafe($data);
    }

    /**
     * Creates a new user
     *
     * If you want the user's account to be activated by default, pass `null` as the callback
     *
     * If you want to make the user verify their email address first, pass an anonymous function as the callback
     *
     * The callback function must have the following signature:
     *
     * `function ($selector, $token)`
     *
     * Both pieces of information must be sent to the user, usually embedded in a link
     *
     * When the user wants to verify their email address as a next step, both pieces will be required again
     *
     * @param bool $requireUniqueUsername whether it must be ensured that the username is unique
     * @param string $email the email address to register
     * @param string $password the password for the new account
     * @param string|null $username (optional) the username that will be displayed
     * @param callable|null $callback (optional) the function that sends the confirmation email to the user
     * @return int the ID of the user that has been created (if any)
     * @throws InvalidEmailException if the email address has been invalid
     * @throws InvalidPasswordException if the password has been invalid
     * @throws UserAlreadyExistsException if a user with the specified email address already exists
     * @throws DuplicateUsernameException if it was specified that the username must be unique while it was *not*
     * @throws AuthError if an internal problem occurred (do *not* catch)
     *
     * @see confirmEmail
     * @see confirmEmailAndSignIn
     */
    protected function createUserInternal($requireUniqueUsername, $email, $password, $username = null, callable $callback = null)
    {
        \ignore_user_abort(true);

        $email = self::validateEmailAddress($email);
        $password = self::validatePassword($password);

        $username = isset($username) ? \trim($username) : null;

        // if the supplied username is the empty string or has consisted of whitespace only
        if ($username === '') {
            // this actually means that there is no username
            $username = null;
        }

        // if the uniqueness of the username is to be ensured
        if ($requireUniqueUsername) {
            // if a username has actually been provided
            if ($username !== null) {
                // count the number of users who do already have that specified username
                $occurrencesOfUsername = $this->db->selectValue(
                    'SELECT COUNT(*) FROM ' . $this->makeTableName(self::$tables['users']) . ' WHERE username = ?',
                    [$username]
                );

                // if any user with that username does already exist
                if ($occurrencesOfUsername > 0) {
                    // cancel the operation and report the violation of this requirement
                    throw new DuplicateUsernameException();
                }
            }
        }

        $password = \password_hash($password, $this->passwordHashAlgo);
        $verified = \is_callable($callback) ? 0 : 1;

        try {
            $this->db->insert(
                $this->makeTableNameComponents(self::$tables['users']),
                [
                    'email' => $email,
                    'password' => $password,
                    'username' => $username,
                    'verified' => $verified,
                    'registered' => \time()
                ]
            );
        }
        // if we have a duplicate entry
        catch (IntegrityConstraintViolationException $e) {
            throw new UserAlreadyExistsException();
        } catch (Error $e) {
            throw new DatabaseError($e->getMessage());
        }

        $newUserId = (int) $this->db->getLastInsertId();

        if ($verified === 0) {
            $this->createConfirmationRequest($newUserId, $email, $callback);
        }

        return $newUserId;
    }

    /**
     * Updates the given user's password by setting it to the new specified password
     *
     * @param int $userId the ID of the user whose password should be updated
     * @param string $newPassword the new password
     * @throws UnknownIdException if no user with the specified ID has been found
     * @throws AuthError if an internal problem occurred (do *not* catch)
     */
    protected function updatePasswordInternal($userId, $newPassword)
    {
        $newPassword = \password_hash($newPassword, $this->passwordHashAlgo);

        try {
            $affected = $this->db->update(
                $this->makeTableNameComponents(self::$tables['users']),
                ['password' => $newPassword],
                ['id' => $userId]
            );

            if ($affected === 0) {
                throw new UnknownIdException();
            }
        } catch (Error $e) {
            throw new DatabaseError($e->getMessage());
        }
    }

    /**
     * Called when a user has successfully logged in
     *
     * This may happen via the standard login, via the "remember me" feature, or due to impersonation by administrators
     *
     * @param int $userId the ID of the user
     * @param string $email the email address of the user
     * @param string $username the display name (if any) of the user
     * @param int $status the status of the user as one of the constants from the {@see Status} class
     * @param int $roles the roles of the user as a bitmask using constants from the {@see Role} class
     * @param int $forceLogout the counter that keeps track of forced logouts that need to be performed in the current session
     * @param bool $remembered whether the user has been remembered (instead of them having authenticated actively)
     * @throws AuthError if an internal problem occurred (do *not* catch)
     */
    protected function onLoginSuccessful($userId, $email, $username, $status, $roles, $forceLogout, $remembered)
    {
        // re-generate the session ID to prevent session fixation attacks (requests a cookie to be written on the client)
        Session::regenerate(true);

        // save the user data in the session variables maintained by this library
        $_SESSION[self::SESSION_FIELD_LOGGED_IN] = true;
        $_SESSION[self::SESSION_FIELD_USER_ID] = (int) $userId;
        $_SESSION[self::SESSION_FIELD_EMAIL] = $email;
        $_SESSION[self::SESSION_FIELD_USERNAME] = $username;
        $_SESSION[self::SESSION_FIELD_STATUS] = (int) $status;
        $_SESSION[self::SESSION_FIELD_ROLES] = (int) $roles;
        $_SESSION[self::SESSION_FIELD_FORCE_LOGOUT] = (int) $forceLogout;
        $_SESSION[self::SESSION_FIELD_REMEMBERED] = $remembered;
        $_SESSION[self::SESSION_FIELD_LAST_RESYNC] = \time();
    }

    /**
     * Returns the requested user data for the account with the specified username (if any)
     *
     * You must never pass untrusted input to the parameter that takes the column list
     *
     * @param string $username the username to look for
     * @param array $requestedColumns the columns to request from the user's record
     * @return array the user data (if an account was found unambiguously)
     * @throws UnknownUsernameException if no user with the specified username has been found
     * @throws AmbiguousUsernameException if multiple users with the specified username have been found
     * @throws AuthError if an internal problem occurred (do *not* catch)
     */
    protected function getUserDataByUsername($username, array $requestedColumns)
    {
        try {
            $projection = \implode(', ', $requestedColumns);

            $users = $this->db->select(
                'SELECT ' . $projection . ' FROM ' . $this->makeTableName(self::$tables['users']) . ' WHERE username = ? LIMIT 2 OFFSET 0',
                [$username]
            );
        } catch (Error $e) {
            throw new DatabaseError($e->getMessage());
        }

        if (empty($users)) {
            throw new UnknownUsernameException();
        } else {
            if (\count($users) === 1) {
                return $users[0];
            } else {
                throw new AmbiguousUsernameException();
            }
        }
    }

    /**
     * Validates an email address
     *
     * @param string $email the email address to validate
     * @return string the sanitized email address
     * @throws InvalidEmailException if the email address has been invalid
     */
    protected static function validateEmailAddress($email)
    {
        if (empty($email)) {
            throw new InvalidEmailException();
        }

        $email = \trim($email);

        if (!\filter_var($email, \FILTER_VALIDATE_EMAIL)) {
            throw new InvalidEmailException();
        }

        return $email;
    }

    /**
     * Validates a password
     *
     * @param string $password the password to validate
     * @return string the sanitized password
     * @throws InvalidPasswordException if the password has been invalid
     */
    protected static function validatePassword($password)
    {
        if (empty($password)) {
            throw new InvalidPasswordException();
        }

        $password = \trim($password);

        if (\strlen($password) < 1) {
            throw new InvalidPasswordException();
        }

        return $password;
    }

    /**
     * Creates a request for email confirmation
     *
     * The callback function must have the following signature:
     *
     * `function ($selector, $token)`
     *
     * Both pieces of information must be sent to the user, usually embedded in a link
     *
     * When the user wants to verify their email address as a next step, both pieces will be required again
     *
     * @param int $userId the user's ID
     * @param string $email the email address to verify
     * @param callable $callback the function that sends the confirmation email to the user
     * @throws AuthError if an internal problem occurred (do *not* catch)
     */
    protected function createConfirmationRequest($userId, $email, callable $callback)
    {
        $selector = self::createRandomString(16);
        $token = self::createRandomString(16);
        $tokenHashed = \password_hash($token, $this->passwordHashAlgo);
        $expires = \time() + 60 * 60 * 24;

        try {
            $this->db->insert(
                $this->makeTableNameComponents(self::$tables['confirmations']),
                [
                    'user_id' => (int) $userId,
                    'email' => $email,
                    'selector' => $selector,
                    'token' => $tokenHashed,
                    'expires' => $expires
                ]
            );
        } catch (Error $e) {
            throw new DatabaseError($e->getMessage());
        }

        if (\is_callable($callback)) {
            $callback($selector, $token);
        } else {
            throw new MissingCallbackError();
        }
    }

    /**
     * Clears an existing directive that keeps the user logged in ("remember me")
     *
     * @param int $userId the ID of the user who shouldn't be kept signed in anymore
     * @param string $selector (optional) the selector which the deletion should be restricted to
     * @throws AuthError if an internal problem occurred (do *not* catch)
     */
    protected function deleteRememberDirectiveForUserById($userId, $selector = null)
    {
        $whereMappings = [];

        if (isset($selector)) {
            $whereMappings['selector'] = (string) $selector;
        }

        $whereMappings['user'] = (int) $userId;

        try {
            $this->db->delete(
                $this->makeTableNameComponents(self::$tables['remembered']),
                $whereMappings
            );
        } catch (Error $e) {
            throw new DatabaseError($e->getMessage());
        }
    }

    /**
     * Triggers a forced logout in all sessions that belong to the specified user
     *
     * @param int $userId the ID of the user to sign out
     * @throws AuthError if an internal problem occurred (do *not* catch)
     */
    protected function forceLogoutForUserById($userId)
    {
        $this->deleteRememberDirectiveForUserById($userId);
        $this->db->exec(
            'UPDATE ' . $this->makeTableName(self::$tables['users']) . ' SET force_logout = force_logout + 1 WHERE id = ?',
            [$userId]
        );
    }

    /**
     * Builds a (qualified) full table name from an optional qualifier, an optional prefix, and the table name itself
     *
     * The optional qualifier may be a database name or a schema name, for example
     *
     * @param string $name the name of the table
     * @return string[] the components of the (qualified) full name of the table
     */
    protected function makeTableNameComponents($name)
    {
        $components = [];

        if (!empty($this->dbSchema)) {
            $components[] = $this->dbSchema;
        }

        if (!empty($name)) {
            if (!empty($this->dbTablePrefix)) {
                $components[] = $this->dbTablePrefix . $name;
            } else {
                $components[] = $name;
            }
        }

        return $components;
    }

    /**
     * Builds a (qualified) full table name from an optional qualifier, an optional prefix, and the table name itself
     *
     * The optional qualifier may be a database name or a schema name, for example
     *
     * @param string $name the name of the table
     * @return string the (qualified) full name of the table
     */
    protected function makeTableName($name)
    {
        $components = $this->makeTableNameComponents($name);

        return \implode('.', $components);
    }
}
